src/tir/schedule/analysis/analysis.cc

```diff
@@ -1376,5 +1376,333 @@ void CheckStorageScope(const ScheduleState& self, String storage_scope) {
+}
+
+bool TensorizeComparator::CompareAnnotation(const std::pair<String, ObjectRef>& lhs,
+                                            const std::pair<String, ObjectRef>& rhs) {
+  if (lhs.first != rhs.first) return false;
+  if (!lhs.second.same_as(rhs.second)) return false;
+  return VisitExpr(Downcast<PrimExpr>(lhs.second), Downcast<PrimExpr>(rhs.second));
+}
+
+bool TensorizeComparator::CompareAnnotationMap(const Map<String, ObjectRef>& lhs,
+                                               const Map<String, ObjectRef>& rhs) {
+  if (lhs.same_as(rhs)) return true;
+  if (lhs.size() != rhs.size()) return false;
+
+  auto sort_map =
+      [](const Map<String, ObjectRef>& map) -> std::vector<std::pair<String, ObjectRef>> {
+    std::vector<std::pair<String, ObjectRef>> ret;
+    ret.reserve(map.size());
+    for (const auto& pair : map) {
+      ret.emplace_back(pair);
+    }
+    sort(ret.begin(), ret.end());
+    return ret;
+  };
+
+  auto lhs_array = sort_map(lhs), rhs_array = sort_map(rhs);
+
+  for (size_t i = 0; i < lhs.size(); ++i) {
+    if (!CompareAnnotation(lhs_array[i], rhs_array[i])) return false;
+  }
+  return true;
+}
+
+bool TensorizeComparator::CompareBuffer(const Buffer& lhs, const Buffer& rhs) {
+  if (lhs.same_as(rhs)) return true;
+  // Remap both buffer itself and buffer data
+  // Skip buffer shape
+  bool equal = DefEqual(lhs, rhs) && DefEqual(lhs->data, rhs->data) &&
+               CompareType(lhs->dtype, rhs->dtype) && lhs.scope() == rhs.scope();
+  if (equal) {
+    rhs_buffer_map_[rhs] = lhs;
+  } else if (assert_mode_) {
+    LOG(FATAL) << "Buffers are not matching between:" << lhs << " and " << rhs;
+  }
+  return equal;
+}
+
+bool TensorizeComparator::CompareBufferRegion(const BufferRegion& lhs, const BufferRegion& rhs) {
+  // Only for block region declaration
+  if (!CompareBuffer(lhs->buffer, rhs->buffer)) return false;
+  // Number of indices in desc_block must be smaller than it in AST
+  if (rhs->region.size() > lhs->region.size()) return false;
+
+  std::vector<Range> lhs_region;
+  for (const auto& range : lhs->region) {
+    lhs_region.push_back(Range::FromMinExtent(range->min, range->extent));
+  }
+  size_t offset = lhs_region.size() - rhs->region.size();
+  // initialize buffer indices
+  bool need_update = false;
+  if (!buffer_indices_.count(lhs->buffer)) {
+    need_update = true;
+    buffer_indices_[lhs->buffer] = std::vector<PrimExpr>();
+  } else {
+    if (offset != buffer_indices_[lhs->buffer].size()) return false;
+  }
+  std::vector<PrimExpr>& indices = buffer_indices_[lhs->buffer];
+  for (size_t i = 0; i < offset; ++i) {
+    const Range& range = lhs_region[i];
+    // High-dim region must be element-wise
+    if (!is_one(range->extent)) return false;
+    if (need_update) {
+      indices.push_back(range->min);
+    } else {
+      // The order matters since we only map inner block_var to outside block_var
+      if (!VisitExpr(range->min, indices[i])) return false;
+    }
+  }
```

Comment:
style:

i could see two different code paths coupled together controlled by `need_update`, which is a bit vague logic. Let's do this instead:

```
if (buffer_indices_.count(lhs->buffer)) {
  for (size_t i = 0; i < offset; ++i) { /* do one thing */ }
} else {
  for (size_t i = 0; i < offset; ++i) { /* do the other thing */ }
}
```